### PR TITLE
select default serial port

### PR DESCRIPTION
**Pull request by haslinghuis**
ℹ️ Working with the configurator it would be nice to have the port automatically detected.

- Added support for detecting `tty` recognition for Linux and OSX to skip the `cu` port.
- Added support for `CP210x` besides `STM`.
- Added sorting of the port list.
- Fixed `dfu` mode.

### CODE REVIEW

**Comment by haslinghuis** [comment]
Also this could be enhanced for OSX to pick the `tty` one instead of `cu`.

File: src/main.html [hunk 1]
```diff
@@ -169,7 +169,7 @@
             <div id="portsinput">
                 <div class="dropdown dropdown-dark">
                     <select class="dropdown-select" id="port" i18n_title="firmwareFlasherManualPort">
-                        <option value="manual">Manual</option>
+                        <option value="manual">Loading</option>
```
Comment (McGiverGim) [inline src/main.html:172]: By default,  the list of ports have the "manual" one. If you want to change it by "Loading" until the list of ports is retrieved is better to change/remove the value too, not only the tag. I will use the localization too to make it translatable.

File: src/js/port_handler.js [hunk 2]
```diff
@@ -200,13 +200,22 @@ PortHandler.update_port_select = function (ports) {
...
     for (let i = 0; i < ports.length; i++) {
         let portText;
+        let opt;
+        let str;
         if (ports[i].displayName) {
             portText = (`${ports[i].path} - ${ports[i].displayName}`);
+            str = `${ports[i].displayName}`;
+            if (str.includes('STM32')) {
+                opt = `<option selected value='${ports[i].path}'>`;
+            } else {
```
Comment (McGiverGim) [inline src/js/port_handler.js:210]: I think the selection of port must be done outside the generation. In this way we can decide if they are several STM32, apply a more complex method, let the generation cleaner, only call it if necessary, etc. 

We usually we use jquery with it, it's cleaner, with a simple: `this.portPickerElement.val(value_to_select).change()` will be enough in both cases.

This is being done at some places of code, for example this is called just after the method you have changed, so better add your logic this:

https://github.com/betaflight/betaflight-configurator/blob/e8b097d6c59a56ffaed6bb7f19490a8bdb85bc92/src/js/port_handler.js#L122-L126

Other place:

https://github.com/betaflight/betaflight-configurator/blob/e8b097d6c59a56ffaed6bb7f19490a8bdb85bc92/src/js/port_handler.js#L92

Other place:

https://github.com/betaflight/betaflight-configurator/blob/e8b097d6c59a56ffaed6bb7f19490a8bdb85bc92/src/js/port_handler.js#L178

You can add it in the if of the first sample, or make a more complex method called from all of this places.

**Review by McGiverGim — CHANGES_REQUESTED** [review]
Two comments, all of this without testing it, so maybe I'm wrong at some place ;)

File: src/js/port_handler.js (hunk 2 above)
Comment (haslinghuis) [reply src/js/port_handler.js:210]: Thanks. Working on it!

Comment (haslinghuis) [reply src/js/port_handler.js]: Line 122 in port_handler.js is already part of finding a new port (after initialization is done or perhaps not yet complete).
Refactored it into a function. I think it's preferred to use last_user_port.

File: src/main.html (hunk 1 above)
Comment (haslinghuis) [reply src/main.html]: Done!

**Comment by haslinghuis** [comment]
Also added support for detecting `tty` recognition for Linux and OSX so it should skip the `/dev/cu.`

File: src/main.html [hunk 3]
```diff
@@ -169,7 +169,7 @@
             <div id="portsinput">
                 <div class="dropdown dropdown-dark">
                     <select class="dropdown-select" id="port" i18n_title="firmwareFlasherManualPort">
-                        <option value="manual">Manual</option>
+                        <option value="loading">Loading</option>
```
Comment (McGiverGim) [inline src/main.html:172]: Add the `Loading` string to the `messages.json` file.

**Review by McGiverGim — CHANGES_REQUESTED** [review]
You forgot the i18n.

The total code of handling of ports is a mesh, I will fix a lot of things, but is outside of this PR. The PR is ok to me.

File: src/main.html (hunk 3 above)
Comment (haslinghuis) [reply src/main.html:172]: Fixed !

**Review by McGiverGim — DISMISSED** [review]

**Comment by haslinghuis** [comment]
Detection worked better on BFC 10.7.0. This function tries to address this by using the new port descriptors but should only correct selection on first init (with USB connected). New port detection should use this function if `!last_used_port` and if multiple ports are found on non Windows to select `tty` (filter out `cu`) but also in the event of plugging in another FC. After that the `last_used_port` should do it's function.

Found out that Radix (which I know is not supported by Betaflight) returns `BrainFPV_RADIX` instead of `STM32`.
`STMElectronics` || `STM32` should be `Betaflight`. So I have to restructure a bit not breaking the new port detection for other products without having to name them.

**Comment by haslinghuis** [comment]
Detection worked better on BFC 10.7.0. This function tries to address this by using the new port descriptors but should only correct selection on first init (with USB connected). New port detection should use this function if `!last_used_port` and if multiple ports are found on non Windows to select `tty` (filter out `cu`) but also in the event of plugging in another FC. After that the `last_used_port` should do it's function.

Found out that Radix (which I know is not supported by Betaflight) returns `BrainFPV_RADIX` instead of `STM32`. Too bad that I cannot find the cause of the initial detection.

Restructured a bit (minimizing) not breaking the new port detection for other products without having to name them.

**Comment by haslinghuis** [comment]
Added sorting of the port list.

**Comment by mikeller** [comment]
Tested on linux:
- works as expected.

Tested on OSX:
- boards with VCP work fine;
- connection not detected any more for boards with CP210x USB to serial converter:

image

**Comment by haslinghuis** [comment]
Thanks for testing! Added `CP210x` to the equation (CP2 because I don't know the full string).

**Comment by haslinghuis** [comment]
Thanks for testing! Added `CP210x` to the comparison (using `CP210` because its a `CP2102` ?)!
Remove last_used_port because it didn't seems to be used.

@mikeller Could you please test again because lacking hardware on this side.

**Review by mikeller — APPROVED** [review]

**Pull request merged**